### PR TITLE
feat(mme): AMF Authentication Client Creation

### PR DESCRIPTION
**Pull request by Lkishor123**
Signed-off-by: Lkishor123 <laawanya.kishor@wavelabs.ai>

## Summary
Changes:
 1. Authentication client creation in AMF.
 2. Message handling from subscriberdb and posting as itti message.

## Test Plan
 1. Unit Testing

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Comment by ulaskozat** [comment]
@Lkishor123 can you update the summary as it looks out of date.

File: lte/gateway/c/core/oai/lib/n11/amf_client_proto_msg_to_itti_msg.cpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+#include "common_types.h"
+#include "security_types.h"
+
+extern "C" {}
+
+using magma::lte::M5GAuthenticationInformationAnswer;
+
+namespace magma5g {
+
+void convert_proto_msg_to_itti_m5g_auth_info_ans(
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/lib/n11/amf_client_proto_msg_to_itti_msg.cpp:33]: can you add test coverage?

File: lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.h [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+using grpc::Status;
+using magma::GRPCReceiver;
+using magma::lte::M5GAuthenticationInformationAnswer;
+using magma::lte::M5GAuthenticationInformationRequest;
+using magma::lte::M5GSubscriberAuthentication;
+
+namespace magma5g {
+
+class M5GAuthenticationServiceClient {
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.h:36]: add test coverage

**Review by ulaskozat — CHANGES_REQUESTED** [review]
Please add code coverage where it is asked by codecov.

File: lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.cpp [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+#include <grpcpp/impl/codegen/client_context.h>
+#include <grpcpp/impl/codegen/status.h>
+#include <cstring>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <thread>
+#include "lte/protos/subscriberauth.grpc.pb.h"
+#include "lte/protos/subscriberauth.pb.h"
+#include "M5GAuthenticationServiceClient.h"
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.cpp]: nit: this should be the top included header

File: lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.cpp [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <google/protobuf/util/time_util.h>
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.cpp]: in general, sort these headers in order of:

1. header file of this implementation file
2. internal C/C++ lib headers
3. oai include headers (3gpp, log)
4. other (grpc, service registry...)

File: lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.cpp [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+M5GAuthenticationInformationRequest create_subs_auth_request(
+    const std::string& imsi, const std::string& snni) {
+  M5GAuthenticationInformationRequest request;
+
+  request.Clear();
+  request.set_user_name(imsi);
+  request.set_serving_network_name(snni);
+
+  return (request);
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.cpp]:
```suggestion
  return request;
```

File: lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.h [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <grpc++/grpc++.h>
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.h]: sort headers accordingly

File: lte/gateway/c/core/oai/lib/n11/amf_client_proto_msg_to_itti_msg.cpp [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+ */
+
+#include <stdint.h>
+#include <string.h>
+#include <iostream>
+#include <string>
+
+#include "3gpp_33.401.h"
+#include "amf_app_messages_types.h"
+#include "amf_client_proto_msg_to_itti_msg.h"
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/lib/n11/amf_client_proto_msg_to_itti_msg.cpp]: move this to top header

**Review by ardzoht — CHANGES_REQUESTED** [review]

File: lte/gateway/c/core/oai/lib/n11/amf_client_proto_msg_to_itti_msg.cpp (hunk 1 above)
Comment (Lkishor123) [reply lte/gateway/c/core/oai/lib/n11/amf_client_proto_msg_to_itti_msg.cpp:33]: Test coverage added in (lte/gateway/c/core/oai/test/n11/test_auth_service_client.cpp)

File: lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.cpp (hunk 5 above)
Comment (Lkishor123) [reply lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.cpp]: Comment addressed.

File: lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.cpp (hunk 4 above)
Comment (Lkishor123) [reply lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.cpp]: Comment addressed.

File: lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.h (hunk 6 above)
Comment (Lkishor123) [reply lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.h]: Comment addressed.

File: lte/gateway/c/core/oai/lib/n11/amf_client_proto_msg_to_itti_msg.cpp (hunk 7 above)
Comment (Lkishor123) [reply lte/gateway/c/core/oai/lib/n11/amf_client_proto_msg_to_itti_msg.cpp]: Comment addressed.

File: lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.h (hunk 2 above)
Comment (Lkishor123) [reply lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.h:36]: Comment addressed.

File: lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.cpp (hunk 3 above)
Comment (Lkishor123) [reply lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.cpp]: Comment addressed.

File: lte/gateway/c/core/oai/common/security_types.h [hunk 8]
```diff
@@ -48,6 +48,13 @@
 #define AUTH_KEY_LENGTH_OCTETS (AUTH_KEY_LENGTH_BITS / 8)
 #define KASME_LENGTH_BITS (256)
 #define KASME_LENGTH_OCTETS (KASME_LENGTH_BITS / 8)
+#define CK_LENGTH_BITS (128)
+#define CK_LENGTH_OCTETS (CK_LENGTH_BITS / 8)
+#define IK_LENGTH_BITS (128)
+#define IK_LENGTH_OCTETS (IK_LENGTH_BITS / 8)
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/common/security_types.h]: do you still need this?

File: lte/gateway/c/core/oai/common/security_types.h [hunk 9]
```diff
@@ -122,8 +129,18 @@ typedef struct eutran_vector_s {
   res_t xres;
   uint8_t autn[AUTN_LENGTH_OCTETS];
   uint8_t kasme[KASME_LENGTH_OCTETS];
+  uint8_t ck[CK_LENGTH_OCTETS];
+  uint8_t ik[IK_LENGTH_OCTETS];
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/common/security_types.h]: I don't think you need these any more...

File: lte/gateway/c/core/oai/test/n11/test_auth_service_client.cpp [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+  authVector1->set_xres_star("xres_star1");
+  authVector1->set_autn("autn1");
+  authVector1->set_kseaf("kseaf1");
+
+  std::cout << "m5gauth_vectors_size :" << response.m5gauth_vectors_size();
+  magma5g::convert_proto_msg_to_itti_m5g_auth_info_ans(
+      response, &amf_app_subs_auth_info_resp_p);
+
+  EXPECT_TRUE(
+      response.m5gauth_vectors_size() ==
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/n11/test_auth_service_client.cpp:48]: You should construct an expected itti message structure with constant values and then compare the returned struct from the function call against it.

**Review by ulaskozat — CHANGES_REQUESTED** [review]
you need to add mock RPC message handlers and use EXPECT_CALL to test the functions. See SessiondMocks.h and sessiond tests for examples)

File: lte/gateway/c/core/oai/test/n11/test_auth_service_client.cpp [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+      amf_app_subs_auth_info_resp_p.auth_info.nb_of_vectors)
+      << "size of response.m5gauth_vectors_size : "
+      << response.m5gauth_vectors_size()
+      << " amf_app_subs_auth_info_resp_p         : "
+      << amf_app_subs_auth_info_resp_p.auth_info.nb_of_vectors;
+}
+
+TEST(test_get_subs_auth_info, get_subs_auth_info) {
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/n11/test_auth_service_client.cpp:56]: this just tests `create_subs_auth_request` not the function `get_subs_auth_info`. You need to create a mock handler library (e.g., check mock handlers in `SessiondMocks.h` and how they are used in test cases via `EXPECT_CALL`).

File: lte/gateway/c/core/oai/lib/n11/amf_client_proto_msg_to_itti_msg.cpp (hunk 1 above)
Comment (ulaskozat) [reply lte/gateway/c/core/oai/lib/n11/amf_client_proto_msg_to_itti_msg.cpp:33]: You need to compare against an expected struct in the testing. What you did is not covering the whole structure.

File: lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.h (hunk 2 above)
Comment (ulaskozat) [reply lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.h:36]: partially. you need to create mock handlers for both `get_subs_auth_info` and `get_subs_auth_info_resync`. As is you are simply testing a function that is called within `get_subs_auth_info`.

File: lte/gateway/c/core/oai/lib/n11/amf_client_proto_msg_to_itti_msg.cpp (hunk 1 above)
Comment (Lkishor123) [reply lte/gateway/c/core/oai/lib/n11/amf_client_proto_msg_to_itti_msg.cpp:33]: The Unit Test will be covered in next PR.

File: lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.h (hunk 2 above)
Comment (Lkishor123) [reply lte/gateway/c/core/oai/lib/n11/M5GAuthenticationServiceClient.h:36]: The Unit Test will be covered in next PR.

File: lte/gateway/c/core/oai/common/security_types.h (hunk 8 above)
Comment (Lkishor123) [reply lte/gateway/c/core/oai/common/security_types.h]: Comment addressed.

File: lte/gateway/c/core/oai/common/security_types.h (hunk 9 above)
Comment (Lkishor123) [reply lte/gateway/c/core/oai/common/security_types.h]: Comment addressed.

File: lte/gateway/c/core/oai/test/n11/test_auth_service_client.cpp (hunk 10 above)
Comment (Lkishor123) [reply lte/gateway/c/core/oai/test/n11/test_auth_service_client.cpp:48]: The Unit Test will be covered in next PR.

File: lte/gateway/c/core/oai/test/n11/test_auth_service_client.cpp (hunk 11 above)
Comment (Lkishor123) [reply lte/gateway/c/core/oai/test/n11/test_auth_service_client.cpp:56]: The Unit Test will be covered in next PR.

**Review by ulaskozat — APPROVED** [review]
LGTM. Additional unit tests to be added in the next PR.

**Review by ardzoht — APPROVED** [review]
lgtm

**Review by themarwhal — APPROVED** [review]

**Pull request merged**